parser resilience

arifogel opened the pull request:
- Fix line splitting
- Fix adaptive prediction loop at EOF
- Don't crash on unrecognized interface names

dhalperi commented:
*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BatfishException.java, line 33 at r1](https://reviewable.io:443/reviews/batfish/batfish/918#-L4nygqG2yekE4PT2H7v:-L4nygqG2yekE4PT2H7w:b-apjjla) ([raw file](https://github.com/batfish/batfish/blob/88cec6c79bb4162bb3b5f407a9ddd4efe62bb189/projects/batfish-common-protocol/src/main/java/org/batfish/common/BatfishException.java#L33)):*
> ```Java
>     public BatfishStackTrace(BatfishException exception) {
>       String stackTrace = ExceptionUtils.getFullStackTrace(exception).replace("\t", "   ");
>       _lines = Arrays.asList(stackTrace.split("\\n", 1));
> ```

what does this do? From Javadoc, seems like it should be able to return at most 1 string, so would not split it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/918)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
Reviewed 20 of 23 files at r1.
Review status: 20 of 23 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Ip6Wildcard.java, line 25 at r1](https://reviewable.io:443/reviews/batfish/batfish/918#-L4o2KMW24rbsc6BqoTh:-L4o2KMW24rbsc6BqoTi:b-qoojqz) ([raw file](https://github.com/batfish/batfish/blob/88cec6c79bb4162bb3b5f407a9ddd4efe62bb189/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Ip6Wildcard.java#L25)):*
> ```Java
>       }
>     } else if (str.contains("/")) {
>       String[] parts = str.split("/", -1);
> ```

I'm not convinced you want all these `-1`s all over the place.

E.g. here, do you really want "::1/" to be recognized as a correct string?

---

*[projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java, line 1901 at r1](https://reviewable.io:443/reviews/batfish/batfish/918#-L4o-iuSCy58PBlSetQ4:-L4o-iuSCy58PBlSetQ5:b-hs9s80) ([raw file](https://github.com/batfish/batfish/blob/88cec6c79bb4162bb3b5f407a9ddd4efe62bb189/projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java#L1901)):*
> ```Java
>               + getLocation(ctx)
>               + getFullText(ctx)
>               + e.getMessage());
> ```

Do we not need any whitespace between things here?

---

*[projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ApplyPathApplicator.java, line 61 at r1](https://reviewable.io:443/reviews/batfish/batfish/918#-L4o2x6lmEYmUodfgaV4:-L4o2x6lmEYmUodfgaV5:bgu6tlh) ([raw file](https://github.com/batfish/batfish/blob/88cec6c79bb4162bb3b5f407a9ddd4efe62bb189/projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ApplyPathApplicator.java#L61)):*
> ```Java
>     String pathQuoted = ctx.path.getText();
>     String pathWithoutQuotes = pathQuoted.substring(1, pathQuoted.length() - 1);
>     String[] pathComponents = pathWithoutQuotes.split("\\s+");
> ```

inconsistent on the `-1`, if you care.

---

*[projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java, line 551 at r1](https://reviewable.io:443/reviews/batfish/batfish/918#-L4nzwgurNsNHJ23yw0l:-L4nzwgurNsNHJ23yw0m:b4d9s4e) ([raw file](https://github.com/batfish/batfish/blob/88cec6c79bb4162bb3b5f407a9ddd4efe62bb189/projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java#L551)):*
> ```Java
>     String iosRecoveryName = "ios-recovery";
>     List<String> configurationNames =
>         ImmutableList.of(iosRecoveryName, "ios-blankish-file", "ios-bad-interface-name");
> ```

Use three separate unit test cases to test the three separate kinds of bugs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/918)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 5 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/common/BatfishException.java, line 33 at r1](https://reviewable.io:443/reviews/batfish/batfish/918#-L4nygqG2yekE4PT2H7v:-L4oHftmdUTUQ0H57BT9:b-886gsx) ([raw file](https://github.com/batfish/batfish/blob/88cec6c79bb4162bb3b5f407a9ddd4efe62bb189/projects/batfish-common-protocol/src/main/java/org/batfish/common/BatfishException.java#L33)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

what does this do? From Javadoc, seems like it should be able to return at most 1 string, so would not split it?
</blockquote></details>

Fixed

---

*[projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java, line 1901 at r1](https://reviewable.io:443/reviews/batfish/batfish/918#-L4o-iuSCy58PBlSetQ4:-L4oIEIeDr_8_tlAQF4E:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/88cec6c79bb4162bb3b5f407a9ddd4efe62bb189/projects/batfish/src/main/java/org/batfish/grammar/cisco/CiscoControlPlaneExtractor.java#L1901)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Do we not need any whitespace between things here?
</blockquote></details>

Done.

---

*[projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ApplyPathApplicator.java, line 61 at r1](https://reviewable.io:443/reviews/batfish/batfish/918#-L4o2x6lmEYmUodfgaV4:-L4oIGElHTBYXywovbud:b-hy3eo) ([raw file](https://github.com/batfish/batfish/blob/88cec6c79bb4162bb3b5f407a9ddd4efe62bb189/projects/batfish/src/main/java/org/batfish/grammar/flatjuniper/ApplyPathApplicator.java#L61)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

inconsistent on the `-1`, if you care.
</blockquote></details>

Intentional, since we don't want to treat whitespace at end of CLI command as an argument

---

*[projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java, line 551 at r1](https://reviewable.io:443/reviews/batfish/batfish/918#-L4nzwgurNsNHJ23yw0l:-L4oKqtYvtU4C2sEfU1b:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/88cec6c79bb4162bb3b5f407a9ddd4efe62bb189/projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java#L551)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Use three separate unit test cases to test the three separate kinds of bugs.
</blockquote></details>

Done.

---

*[projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Ip6Wildcard.java, line 25 at r1](https://reviewable.io:443/reviews/batfish/batfish/918#-L4o2KMW24rbsc6BqoTh:-L4oHhXYebRnmKxNl1Rk:bwvyox4) ([raw file](https://github.com/batfish/batfish/blob/88cec6c79bb4162bb3b5f407a9ddd4efe62bb189/projects/batfish-common-protocol/src/main/java/org/batfish/datamodel/Ip6Wildcard.java#L25)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

I'm not convinced you want all these `-1`s all over the place.

E.g. here, do you really want "::1/" to be recognized as a correct string?
</blockquote></details>

Reverted

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/918)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 23 files at r1, 4 of 17 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java, line 581 at r2](https://reviewable.io:443/reviews/batfish/batfish/918#-L4oMFefxsodWuBJjmgj:-L4oMFefxsodWuBJjmgk:bwp289l) ([raw file](https://github.com/batfish/batfish/blob/e03f20e50856a568aee0d77cc6162b843516fd87/projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java#L581)):*
> ```Java
>     assertThat(new InterfaceAddress("10.0.0.4/32"), isIn(l4Prefixes));
>     assertThat(configurations, hasKey("ios-bad-interface-name"));
>     assertThat(configurations.entrySet(), hasSize(3));
> ```

suspect this change needs reverting now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/918#-:-L4oMDYPcVeMcMR1BGsn:bnfp4nl)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Review status: 9 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java, line 581 at r2](https://reviewable.io:443/reviews/batfish/batfish/918#-L4oMFefxsodWuBJjmgj:-L4oNftZj6ou6sWCcGhg:b-896fix) ([raw file](https://github.com/batfish/batfish/blob/e03f20e50856a568aee0d77cc6162b843516fd87/projects/batfish/src/test/java/org/batfish/grammar/cisco/CiscoGrammarTest.java#L581)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

suspect this change needs reverting now.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/918)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/918)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/918)*
<!-- Sent from Reviewable.io -->

arifogel commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/918)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/918#-:-L4qNsGjwiXxvIoXkvj1:bnfp4nl)*
<!-- Sent from Reviewable.io -->
